Fix18806

bendk opened the pull request:
Bunch of fixes for #18806.  I'd love review, but please don't merge until Janet says that its fast enough on her machine.

Commit:
Fix bug when removing items then re-adding them.

We were pushing things to the echonest queue wrong, So that the items were
((path, None), ()) instead of (path, None) like they should have been.
By accident pop() still returned the right values so things seemed to work.

However, remove_tasks_from_path() would not work properly and we would
sometimes end up with a string of warnings and possibly errors.  I could get
this when I clicked back and forth between the bulk run/remove online lookup
buttons.

Commit:
bz18806: speedup fetching MetadataStatus objects in set_net_lookup_enabled()

Before we were fetching each one individually by path, even if we were
changing the value globally.  This speeds up that case by making a view that
contains all the MetadataStatus objects that would need to be changed.  This
is most noticable when MetadataStatus objects haven't been loaded yet and put
into the cache.

Commit:
bz18806: call get_metadata() lazily for echonest.

If we add a bunch of paths to our echonest processor, then we were calling
get_metadata() on all of them.  If the user clicks the buttons in the
preferences window, this could be thousands of paths.  get_metadata() is too
slow to be called on that many paths at once.  Instead, try to wait to call it
until we really need the data.

This was probably the slowest part of set_net_lookup_enabled()

paulswartz commented on tv/lib/metadata.py:
```diff
@@ -698,6 +713,38 @@ def _should_process_codegen_queue(self):
+            try:
+                metadata = metadata_fetcher()
+            except KeyError:
+                logging.warn("_process_metadata_fetch_queue: metadata_fetcher "
```
Is this log item going to be useful? Some things I can think to include:
- item
- fetcher function
- traceback

Also, what about other exceptions?

bendk replied:
Yeah, all of those things would be good.  By item, I'm going to use path.

I was thinking that other exceptions should be propagated upward, but now I don't think so anymore.  I'll change to catch StandardError.

Commit:
Better logging if metadata_fetcher raises an exception.

Also added a comment and an explicit continue statement to make it more clear
what we do next.

bendk commented:
Also, janet checked out the branch and it was running fast enough for her.  So this can be merged once it passes review.

paulswartz commented on tv/lib/metadata.py:
```diff
@@ -1471,23 +1527,25 @@ def _run_movie_data(self, path):
-            except KeyError:
-                pass
-        self.echonest_processor.add_path(path, echonest_metadata)
+        def metadata_fetcher():
```
Do we still need this comment, since we're doing it later to avoid the slowdown?

Commit:
Fixed comments.

Removed comment aboing things being slow.  Now that we're doing it lazily that
shouldn't be an issue.  Reformatted other ones to fix in 78char width.

paulswartz commented:
Looks good!
